src/metabase/api/database.clj

```diff
@@ -352,6 +353,49 @@
                                      secret/expand-db-details-inferred-secret-values
                                      (assoc :can-manage true)))))
 
+(defmulti ^:private database-usage-info
+  {:arglists '([model database-id table-ids])}
+  (fn [model _database-id _table-ids] (keyword model)))
+
+(defmethod database-usage-info :question
+  [_ db-id _table-ids]
+  {:select [[:%count.* :question]]
+   :from   [:report_card]
+   :where  [:and
+            [:= :database_id db-id]
+            [:= :dataset false]]})
+
+(defmethod database-usage-info :dataset
+  [_ db-id _table-ids]
+  {:select [[:%count.* :dataset]]
+   :from   [:report_card]
+   :where  [:and
+            [:= :database_id db-id]
+            [:= :dataset true]]})
+
+(defmethod database-usage-info :metric
+  [_ _db-id table-ids]
+  {:select [[:%count.* :metric]]
+   :from   [:metric]
+   :where  [:in :table_id table-ids]})
+
+(defmethod database-usage-info :segment
+  [_ _db-id table-ids]
+  {:select [[:%count.* :segment]]
+   :from   [:segment]
+   :where  [:in :table_id table-ids]})
```

Comment:
Two things:

1. I think it makes more sense to have this in the model, not the API ns.

2. I'd call it `database-usage-query` or something.

Reply:
I think our general convention is we only put things into models if it meant to be used in multiple places. Granting and removing permissions is a good example.

Database usage will probably be used only in database API, so I think it's okay if we leave them there.

One [example](https://github.com/metabase/metabase/blob/5e81263604681f9d3256d66ac6c04a6349f918cd/src/metabase/api/collection.clj#L154) is the logic to fetch items for a collection, it's a lot of logic but we still leave them in API ns.

> I'd call it database-usage-query or something.

totally. this was in draft so I just whipped up code as fast as I could. I was going to tailor it later.